File: apps/server-mac/PrismServer/Services/PairingCodeService.swift
import Foundation

struct DisplayPairingCode: Equatable {
    let code: String
    let expiresAt: Date

    var expirationSummary: String {
        expiresAt.formatted(date: .omitted, time: .shortened)
    }
}

final class PairingCodeService {
    private struct Response: Decodable {
        struct PairingCodePayload: Decodable {
            let code: String
            let expiresAt: Date
        }

        let ok: Bool
        let pairingCode: PairingCodePayload
    }

    func createPairingCode(apiPort: Int) async throws -> DisplayPairingCode {
        guard let url = URL(string: "http://127.0.0.1:\(apiPort)/api/local/pairing/codes") else {
            throw PairingCodeError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 5

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 201 else {
            throw PairingCodeError.requestFailed("Could not generate a pairing code. Make sure Prism Server is running.")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let decoded = try decoder.decode(Response.self, from: data)
        guard decoded.ok else {
            throw PairingCodeError.requestFailed("Prism Server did not accept the pairing request.")
        }

        return DisplayPairingCode(
            code: decoded.pairingCode.code,
            expiresAt: decoded.pairingCode.expiresAt
        )
    }
}

enum PairingCodeError: LocalizedError {
    case invalidURL
    case requestFailed(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Could not build the local pairing URL."
        case .requestFailed(let message):
            return message
        }
    }
}

File: apps/server-windows/core/Services/PairingCodeService.cs
using System.Net.Http.Json;
using PrismServer.Core.Models;

namespace PrismServer.Core.Services;

public sealed class PairingCodeService
{
    private readonly HttpClient _httpClient;

    public PairingCodeService(HttpClient httpClient)
    {
        _httpClient = httpClient;
    }

    public async Task<DisplayPairingCode> CreatePairingCodeAsync(int apiPort, CancellationToken cancellationToken = default)
    {
        var uri = new Uri($"http://127.0.0.1:{apiPort}/api/local/pairing/codes");
        using var request = new HttpRequestMessage(HttpMethod.Post, uri);
        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        cts.CancelAfter(TimeSpan.FromSeconds(5));

        using var response = await _httpClient.SendAsync(request, cts.Token).ConfigureAwait(false);
        if ((int)response.StatusCode != 201)
        {
            throw new InvalidOperationException("Could not generate a pairing code. Make sure Prism Server is running.");
        }

        var decoded = await response.Content.ReadFromJsonAsync<PairingResponse>(cancellationToken: cts.Token).ConfigureAwait(false);
        if (decoded is null || !decoded.Ok || decoded.PairingCode is null)
        {
            throw new InvalidOperationException("Prism Server did not accept the pairing request.");
        }

        return new DisplayPairingCode(decoded.PairingCode.Code, decoded.PairingCode.ExpiresAt);
    }

    private sealed record PairingResponse(bool Ok, PairingCodePayload? PairingCode);
    private sealed record PairingCodePayload(string Code, DateTimeOffset ExpiresAt);
}
